### PR TITLE
Egress support v2

### PR DESCRIPTION
**Pull request by RomanDzhabarov**
Work is in progress for this item. Need to add tests etc.

### CODE REVIEW

File: source/common/http/conn_manager_impl.h [hunk 1]
```diff
@@ -161,7 +161,12 @@ class ConnectionManagerConfig {
   /**
    * @return true if tracing is enabled otherwise it returns false;
    */
-  virtual bool isTracing() PURE;
+  virtual bool isTracing(const Http::AccessLog::RequestInfo& request_info) PURE;
```
Comment (mattklein123) [inline source/common/http/conn_manager_impl.h]: I would rename shouldTraceRequest(...)

File: include/envoy/tracing/http_tracer.h [hunk 2]
```diff
@@ -6,6 +6,30 @@
...
+  // Trace only when there is an upstream failure reason.
+  UpstreamFailureReason
+};
+
+/**
+ * Configuration for tracing which is set on the connection manager level.
+ * Http Tracing can be enabled/disabled on a per connection manager basis.
+ * Here we specify some specific for connection manager settings.
+ */
+struct TracingConnectionManagerConfig {
```
Comment (RomanDzhabarov) [inline include/envoy/tracing/http_tracer.h]: right, moving that out.

File: source/common/http/conn_manager_utility.cc [hunk 3]
```diff
@@ -131,4 +131,22 @@ void ConnectionManagerUtility::mutateResponseHeaders(Http::HeaderMap& response_h
...
+
+  switch (config.value().tracing_type_) {
+  case Http::TracingType::All:
+    return true;
+  case Http::TracingType::UpstreamFailureReason:
+    return request_info.failureReason() != Http::AccessLog::FailureReason::None;
+  }
+
+  // This should not happen as switch above covers all cases, this is just to make compiler happy.
+  throw EnvoyException("unknown tracing type");
```
Comment (ccaraman) [inline source/common/http/conn_manager_utility.cc]: Can you add the tracing_type value to the exception?

Comment (RomanDzhabarov) [inline source/common/http/conn_manager_utility.cc]: added

File: docs/configuration/http_conn_man/http_conn_man.rst [hunk 4]
```diff
@@ -71,11 +71,34 @@ add_user_agent
...
   documentation for more information. Defaults to false.
 
-.. _config_http_conn_man_tracing_enabled:
-
-tracing_enabled
-  *(optional, boolean)* Whether the connection manager emits :ref:`tracing <arch_overview_tracing>`
-  data to the :ref:`configured tracing provider <config_tracing>`. Defaults to false.
+.. _config_http_conn_man_tracing:
+
+tracing
```
Comment (mattklein123) [inline docs/configuration/http_conn_man/http_conn_man.rst]: generally a sub-level json object gets its own section (see other docs). Can you make a new section on this page, put the nested json docs there, and link to it from here (again see other docs)

File: include/envoy/tracing/http_tracer.h [hunk 5]
```diff
@@ -6,6 +6,13 @@
 
 namespace Tracing {
 
+class TracingContext {
```
Comment (mattklein123) [inline include/envoy/tracing/http_tracer.h]: docs

File: source/common/http/conn_manager_impl.h [hunk 6]
```diff
@@ -76,6 +76,23 @@ struct ConnectionManagerStats {
   Stats::Store& store_;
 };
 
+enum class TracingType {
+  // Trace all traceable requests.
+  All,
+  // Trace only when there is an upstream failure reason.
+  UpstreamFailureReason
```
Comment (mattklein123) [inline source/common/http/conn_manager_impl.h]: I would call this UpstreamFailure

File: source/common/http/conn_manager_utility.cc (hunk 3 above)
Comment (mattklein123) [inline source/common/http/conn_manager_utility.cc]: Just use NOT_IMPLEMENTED or something clear that it should not happen

File: source/common/http/conn_manager_impl.h (hunk 6 above)
Comment (RomanDzhabarov) [inline source/common/http/conn_manager_impl.h]: fixed

File: include/envoy/tracing/http_tracer.h (hunk 5 above)
Comment (RomanDzhabarov) [inline include/envoy/tracing/http_tracer.h]: fixed

File: source/common/http/conn_manager_utility.cc (hunk 3 above)
Comment (RomanDzhabarov) [inline source/common/http/conn_manager_utility.cc]: i'll switch to NOT_IMPLEMENTED here and refine the comment to emphasize that compiler also enforces all types covered by switch above.

**Pull request merged**